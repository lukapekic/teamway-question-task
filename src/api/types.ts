interface Answer {
  id: number;
  label: string;
}

export interface Question {
  id: number;
  question: string;
  answers: Answer[];
}

export enum PersonalityType {
  INTEROVER = "Introvert",
  EXTROVERT = "Extrovert",
}
