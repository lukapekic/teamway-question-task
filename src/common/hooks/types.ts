export interface Answers {
  [key: string]: number;
}
