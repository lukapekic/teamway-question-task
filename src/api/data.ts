import { Question } from "./types";

export const questions: Question[] = [
  {
    id: 1,
    question:
      "You're really busy at work and a colleague is telling you their life story and personal woes. You:",
    answers: [
      {
        id: 1,
        label: "Don't dare to interrupt them",
      },
      {
        id: 2,
        label:
          "Think it's more important to give them some of your time; work can wait",
      },
      {
        id: 3,
        label: "Listen, but with only with half an ear",
      },
      {
        id: 4,
        label: "Interrupt and explain that you are really busy at the moment",
      },
    ],
  },
  {
    id: 2,
    question:
      "You've been sitting in the doctor's waiting room for more than 25 minutes. You:",
    answers: [
      {
        id: 1,
        label: "Look at your watch every two minutes",
      },
      {
        id: 2,
        label: "Bubble with inner anger, but keep quiet",
      },
      {
        id: 3,
        label:
          "Explain to other equally impatient people in the room that the doctor is always running late",
      },
      {
        id: 4,
        label: "Complain in a loud voice, while tapping your foot impatiently",
      },
    ],
  },
  {
    id: 3,
    question: "You are taking part in a guided tour of a museum. You:",
    answers: [
      {
        id: 1,
        label:
          "Are a bit too far towards the back so don't really hear what the guide is saying",
      },
      {
        id: 2,
        label: "Follow the group without question",
      },
      {
        id: 3,
        label: "Make sure that everyone is able to hear properly",
      },
      {
        id: 4,
        label:
          "Are right up the front, adding your own comments in a loud voice",
      },
    ],
  },
  {
    id: 4,
    question: "This morning, your agenda seems to be free. You:",
    answers: [
      {
        id: 1,
        label: "Know that somebody will find a reason to come and bother you",
      },
      {
        id: 2,
        label:
          "Heave a sigh of relief and look forward to a day without stress",
      },
      {
        id: 3,
        label:
          "Question your colleagues about a project that's been worrying you",
      },
      {
        id: 4,
        label:
          "Pick up the phone and start filling up your agenda with meetings",
      },
    ],
  },
  {
    id: 5,
    question:
      "You're out with a group of friends and there's a person who's quite shy and doesn't talk much. You:",
    answers: [
      {
        id: 1,
        label: "Notice that they're alone, but don't go over to talk with them",
      },
      {
        id: 2,
        label: "Go and have a chat with them",
      },
      {
        id: 3,
        label: "Shoot some friendly smiles in their direction",
      },
      {
        id: 4,
        label: "Hardly notice them at all",
      },
    ],
  },
];
