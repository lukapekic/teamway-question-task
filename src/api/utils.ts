import { Answers } from "../common/hooks/types";
import { PersonalityType } from "./types";

const personalityAnswers = [
  PersonalityType.INTEROVER,
  PersonalityType.EXTROVERT,
  PersonalityType.EXTROVERT,
  PersonalityType.INTEROVER,
];

export const getPersonalityType = (answers: Answers): PersonalityType => {
  const averageAnswerNumber = Object.entries(answers).reduce(
    (previousValue, currentValue) => {
      const maxOccurencies = Math.max(previousValue[1], currentValue[1]);
      return previousValue.includes(maxOccurencies)
        ? previousValue
        : currentValue;
    },
    ["0", 0]
  )[0];

  return personalityAnswers[Number(averageAnswerNumber)];
};
