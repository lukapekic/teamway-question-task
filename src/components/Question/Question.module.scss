@use "src/common/styles/colors" as c;
@use "src/common/styles/fonts" as f;

.question {
    opacity: 1;
    transition: opacity 1s ease-in-out;

    &--disabled {
        display: none;
        opacity: 0;
    }
}

.answers {
    &__wrapper {
        border-radius: 0.25rem;
        padding: 1.5rem;
        background-color: rgba(c.$tertiary-background, 0.25);
    }

    &__item {
        list-style: none;
        padding: 1rem;
        margin-bottom: 1rem;
        
        color: c.$primary-text;
        font-family: f.$font-secondary;
        
        font-size: 0.75rem;
        cursor: pointer;
        border-radius: 0.25rem;
        background-color: rgba(c.$secondary-background, 0.25);
        transition: background-color 0.3s ease-in-out;
        

        &--active {
            background-color: c.$secondary-background;
        }
    }
}