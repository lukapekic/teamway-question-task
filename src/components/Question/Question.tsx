import classNames from "classnames";
import { FunctionComponent } from "react";
import type { Question as QuestionType } from "../../api/types";
import style from "./Question.module.scss";

interface QuestionProps extends QuestionType {
  onClick: (answerId: number) => void;
  active?: boolean;
  selectedAnswer: null | number;
}

export const Question: FunctionComponent<QuestionProps> = ({
  question,
  onClick,
  answers,
  active,
  selectedAnswer,
}) => {
  return (
    <div
      className={classNames({
        [style["question"]]: true,
        [style["question--disabled"]]: !active,
      })}
    >
      <h1>{question}</h1>
      <div className={style["answers__wrapper"]}>
        {answers.map(({ id, label }) => {
          return (
            <li
              data-testid={"questionComponent-answer" + id}
              key={id}
              onClick={() => onClick(id)}
              className={classNames({
                [style["answers__item"]]: true,
                [style["answers__item--active"]]: selectedAnswer === id,
              })}
            >
              {id} - {label}
            </li>
          );
        })}
      </div>
    </div>
  );
};
