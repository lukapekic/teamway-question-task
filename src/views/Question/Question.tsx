import type { FormEvent, FunctionComponent } from "react";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useQuestions } from "../../common/hooks";
import { Button, ButtonType } from "../../components";
import { Question } from "../../components";
import style from "./Question.module.scss";

export const QuestionPageView: FunctionComponent = () => {
  const navigate = useNavigate();
  const [selectedAnswer, setSelectedAnswer] = useState<number | null>(null);
  const { questions, submitResult, activeQuestion } = useQuestions((result) =>
    navigate("/result", { state: result })
  );

  return (
    <div className={style["question__wrapper"]}>
      <form
        data-testid="questionPage-form"
        onSubmit={(event: FormEvent) => {
          event.preventDefault();
          submitResult(selectedAnswer as number);
          setSelectedAnswer(null);
        }}
      >
        {questions.length !== 0 &&
          questions.map(({ question, answers, id }, index) => (
            <Question
              key={id}
              id={id}
              question={question}
              answers={answers}
              onClick={setSelectedAnswer}
              active={activeQuestion === index}
              selectedAnswer={selectedAnswer}
            />
          ))}
        <div className={style["action__wrapper"]}>
          <Button
            data-testid="questionPage-cancelButton"
            type="reset"
            variant={ButtonType.SECONDARY}
            onClick={() => navigate("/")}
          >
            Cancel
          </Button>
          <Button
            data-testid="questionPage-submitButton"
            type="submit"
            disabled={!selectedAnswer}
          >
            Next
          </Button>
        </div>
      </form>
    </div>
  );
};
