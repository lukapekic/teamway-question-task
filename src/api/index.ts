import { Answers } from "../common/hooks/types";
import { questions } from "./data";
import { getPersonalityType } from "./utils";

export const getQuestions = () => Promise.resolve([...questions]);
export const submitQuestions = (answers: Answers) => {
  const personality = getPersonalityType(answers);
  return Promise.resolve(
    new Response(JSON.stringify(personality), { status: 200 })
  );
};
