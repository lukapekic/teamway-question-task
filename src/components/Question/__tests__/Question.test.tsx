import { fireEvent, render } from "@testing-library/react";
import { Question as QuestionType } from "../../../api/types";
import { Question } from "../Question";

const question: QuestionType = {
  id: 1,
  question: "Question 1",
  answers: [
    {
      id: 1,
      label: "Answer 1",
    },
    {
      id: 2,
      label: "Answer 2",
    },
  ],
};

describe("QuestionComponent", () => {
  const handleClick = jest.fn();

  const setupComponent = () => {
    const component = render(
      <Question
        question={question.question}
        answers={question.answers}
        id={question.id}
        selectedAnswer={1}
        onClick={handleClick}
      />
    );

    return { component };
  };

  it("should render successfuly", () => {
    const { component } = setupComponent();
    expect(component).toMatchSnapshot();
  });

  it("should call onClick prop when answer is clicked", async () => {
    const { component } = setupComponent();

    const firstAnswer = await component.findByTestId(
      "questionComponent-answer1"
    );
    fireEvent.click(firstAnswer);

    expect(handleClick).toHaveBeenCalledWith(1);
  });
});
