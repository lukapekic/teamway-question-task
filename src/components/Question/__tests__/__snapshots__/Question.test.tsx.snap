// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionComponent should render successfuly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="question question--disabled"
      >
        <h1>
          Question 1
        </h1>
        <div
          class="answers__wrapper"
        >
          <li
            class="answers__item answers__item--active"
            data-testid="questionComponent-answer1"
          >
            1
             - 
            Answer 1
          </li>
          <li
            class="answers__item"
            data-testid="questionComponent-answer2"
          >
            2
             - 
            Answer 2
          </li>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="question question--disabled"
    >
      <h1>
        Question 1
      </h1>
      <div
        class="answers__wrapper"
      >
        <li
          class="answers__item answers__item--active"
          data-testid="questionComponent-answer1"
        >
          1
           - 
          Answer 1
        </li>
        <li
          class="answers__item"
          data-testid="questionComponent-answer2"
        >
          2
           - 
          Answer 2
        </li>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
